langlib/lang.value/src/main/java/org/ballerinalang/langlib/value/FromJsonWithType.java

```diff
@@ -0,0 +1,268 @@
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+package org.ballerinalang.langlib.value;
+
+import org.ballerinalang.jvm.BallerinaErrors;
+import org.ballerinalang.jvm.BallerinaValues;
+import org.ballerinalang.jvm.StringUtils;
+import org.ballerinalang.jvm.TypeChecker;
+import org.ballerinalang.jvm.TypeConverter;
+import org.ballerinalang.jvm.XMLFactory;
+import org.ballerinalang.jvm.commons.TypeValuePair;
+import org.ballerinalang.jvm.scheduling.Strand;
+import org.ballerinalang.jvm.types.BArrayType;
+import org.ballerinalang.jvm.types.BField;
+import org.ballerinalang.jvm.types.BMapType;
+import org.ballerinalang.jvm.types.BRecordType;
+import org.ballerinalang.jvm.types.BTableType;
+import org.ballerinalang.jvm.types.BTupleType;
+import org.ballerinalang.jvm.types.BType;
+import org.ballerinalang.jvm.types.BTypes;
+import org.ballerinalang.jvm.types.TypeTags;
+import org.ballerinalang.jvm.util.exceptions.BLangExceptionHelper;
+import org.ballerinalang.jvm.util.exceptions.BallerinaException;
+import org.ballerinalang.jvm.util.exceptions.RuntimeErrors;
+import org.ballerinalang.jvm.values.ArrayValue;
+import org.ballerinalang.jvm.values.ArrayValueImpl;
+import org.ballerinalang.jvm.values.ErrorValue;
+import org.ballerinalang.jvm.values.MapValue;
+import org.ballerinalang.jvm.values.MapValueImpl;
+import org.ballerinalang.jvm.values.RefValue;
+import org.ballerinalang.jvm.values.StringValue;
+import org.ballerinalang.jvm.values.TableValueImpl;
+import org.ballerinalang.jvm.values.TupleValueImpl;
+import org.ballerinalang.jvm.values.TypedescValue;
+import org.ballerinalang.jvm.values.api.BString;
+import org.ballerinalang.natives.annotations.BallerinaFunction;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.ballerinalang.jvm.BallerinaErrors.createError;
+import static org.ballerinalang.jvm.util.exceptions.BallerinaErrorReasons.VALUE_LANG_LIB_CONVERSION_ERROR;
+import static org.ballerinalang.jvm.util.exceptions.BallerinaErrorReasons.VALUE_LANG_LIB_CYCLIC_VALUE_REFERENCE_ERROR;
+import static org.ballerinalang.jvm.util.exceptions.RuntimeErrors.INCOMPATIBLE_CONVERT_OPERATION;
+import static org.ballerinalang.util.BLangCompilerConstants.VALUE_VERSION;
+
+/**
+ * Extern function lang.values:fromJsonWithType.
+ *
+ * @since 2.0
+ */
+@BallerinaFunction(
+        orgName = "ballerina",
+        packageName = "lang.value", version = VALUE_VERSION,
+        functionName = "fromJsonWithType",
+        isPublic = true
+)
+public class FromJsonWithType {
+    private static final String AMBIGUOUS_TARGET = "ambiguous target type";
+
+    public static Object fromJsonWithType(Strand strand, Object v, TypedescValue t) {
+        BType describingType = t.getDescribingType();
+        try {
+            return convert(v, describingType, new ArrayList<>(), t, strand);
```

Comment:
Is there a need to pass both `t` and `describingType`?

Reply:
I followed the same implementation as in the `cloneWithType`. 

`convert` method is get called later. Sometimes, `t` and `targetType` both are  needed

Reply:
My point was if you have `t`, you can always get the type from it. Anyways, not a major issue. Just gives the impression that target type can be some other type than what's described by `t`